fms/models/__init__.py

```diff
@@ -82,4 +87,157 @@ def get_model(
+
+    model = FSDP(
+        model,
+        param_init_fn=init_fn,
+        sync_module_states=sync_module_states,
+        device_id=local_rank,
+        limit_all_gathers=True,
+        use_orig_params=True,
+        # TODO: add wrap and mixed precision policies.
+        # auto_wrap_policy=wrapping_policy,
+        # mixed_precision=mp_policy,
+        sharding_strategy=dp_strategy,
+    )
+    # TODO: add activation checkpointing.
+    return model
+
+
+def _is_dp(distributed_strategy):
+    return distributed_strategy in {"fsdp", "hsdp", "ddp"}
+
+
+def load_model(
+    architecture: str,
+    variant: str,
+    model_path: str = None,
+    source: str = None,
+    device_type: str = "cpu",
+    distributed_strategy: str = None,
+    checkpoint_sharding: str = None,
+    group: ProcessGroup = None,
+    **kwargs,
+):
+    """
+    Load an instance of a model with weights.
+
+    Args:
+    architecture: the model architecture, e.g. llama. See
+                `models.list_models()`.
+    variant: the configuration of the model, e.g. 7b. See
+                `models.list_variants(architecture)`
+    model_path: the path to the state_dict of weights
+    device_type: where to load the model
+    distributed_strategy: None, 'fsdp', 'hsdp', 'tp', or 'mp'.
+    checkpoint_sharding: how the checkpoint files are sharded: None, 'tp',
+                'fsdp', or 'layer'. If None, guess based on files.
+    source: If the weights in the state dict didn't come from an FMS model,
+                `source` specifies which conversion function might be needed.
+                See `serialization.list_sources(architecture)`
+    group: ProcessGroup
+    """
+    local_rank, world_size = distributed.rank_and_world(group)
+
+    if distributed_strategy is None or distributed_strategy == "":
+        if world_size > 1:
+            distributed_strategy = "tp"
+        elif torch.cuda.device_count() > 1:
+            distributed_strategy = "mp"
+
+    device = torch.device(device_type, local_rank)
+    initial_device = device
+    if (
+        _is_dp(distributed_strategy)
+        and local_rank != 0
+        and checkpoint_sharding != "fsdp"
+    ):
+        initial_device = "meta"
+
+    fms_sd = serialization.load_state_dict(
+        model_path,
+        distributed_strategy,
+        checkpoint_sharding,
+        initial_device,
+        local_rank,
+        world_size,
+    )
+    fms_sd = serialization.get_adapted(architecture, source, fms_sd)
+
+    extra_args = kwargs
+    if distributed_strategy == "tp":
+        print("using tensor parallel")
+        extra_args["distributed_strategy"] = TensorParallelStrategy()
+    elif distributed_strategy == "mp":
+        print("using model parallel")
+        devices = [i for i in range(torch.cuda.device_count())]
+        extra_args["distributed_strategy"] = UniformModelParallelStrategy(
+            devices, _guess_num_layers(fms_sd)
```

Comment:
there should be a way to maybe give a hint to _guess_num_layers or give an absolute number of layers if known

Reply:
I'd like to avoid polluting the API for a rare case. How about if I just don't apply a distribution strategy if there's already one in `extra_args`? then a user can specify one manually if they like, bipassing this code.

Reply:
made this change (now if user specifies in extra_args we don't override)